listener.go

```diff
@@ -439,6 +439,7 @@ func (l *listener) loadCert(currentConn net.Conn) (*tls.Certificate, error) {
 	}
 
 	l.cert = &cert
+	close(l.certIsSetSig)
```

Comment:
This feels it may be an improper way to use a channel. We are never actually passing anything into it so what we really want is probably just a mutex.

Reply:
We are sending information that the cert has been populated across routines which I believe is the correct way to use channels. We could send an empty struct on the channel, but closing the channel allows everyone waiting and successive calls to be unblocked. 
IMO a mutex is better suited for stopping multiple threads/routines from accessing the same data at the same time, which is not really what we are trying to accomplish. We are more interested in knowing when the cert is ready. I believe this use case is similar to how channels are used in [time.After](https://pkg.go.dev/time#After). 
All that being said, we can accomplish our goal with either implementation, and neither is inherently better than the other.

Reply:
I think using channels here makes sense, however I would much rather merge this for 2.6.7: https://github.com/rancher/dynamiclistener/pull/64
And then come back to this at the beginning of 2.6.8. We know that sleeping is a valid solution. If this PR misses a spot that might load the cert, and then that fails to close the channel we could easily introduce bugs. But in 2.6.8 I'm fine with this solution.

Reply:
With a channel, do you need to use `select` to add process of `ctx.Done()` ?

In my ENV, here it never meet `<-l.certIsSetSig` due to `if !cert.IsValidTLSSecret(secret) ` in `loadCert`,  and it looks necessary to add `error/warn log` before `if !cert.IsValidTLSSecret(secret) ` returns.

```
			select {
			case <-ctx.Done():
				return
                        case <-l.certIsSetSig: ...
```

Reply:
We do this (block on reading from a channel that is only closed, never written to) all over the place in other rancher projects. It's a solid pattern.

Reply:
but you never meet deadloop before,  now the deadloop comes

Reply:
Yeah I suspect that the channel is not getting closed on all code paths, which means that either:
* there are other paths to setting `l.cert` that aren't closing the channel
* the channel in the struct is being set too late, and is changed after the goroutine blocks on the read

Reply:
@w13915984028 I like the suggestion of adding the context in the select so we can kill the routine with a cancel.
Before this change, instead of a deadlock we had a spinlock, which behaves the same, except the spin lock uses CPU time. @brandond, could you double-check me? But I believe we can guarantee both those scenarios you suggested do not occur because we are dealing with a private struct and member variable.